src/Datadog.Trace/DiagnosticListeners/AspNetCoreDiagnosticObserver.cs

```diff
@@ -304,29 +534,253 @@ private void OnMvcBeforeAction(object arg)
+                    var url = GetUrl(httpContext.Request);
+                    if (!string.Equals(url, trackingFeature.Url))
+                    {
+                        // URL has changed from original, so treat this execution as a "subsequent" request
+                        // Typically occurs for 404s for example
+                        isFirstExecution = false;
+                    }
+                }
+
                 ActionDescriptor actionDescriptor = typedArg.ActionDescriptor;
-                HttpRequest request = typedArg.HttpContext.Request;
+                HttpRequest request = httpContext.Request;
+                IDictionary<string, string> routeValues = actionDescriptor.RouteValues;
+
+                string controllerName = routeValues.TryGetValue("controller", out controllerName)
+                    ? controllerName?.ToLowerInvariant()
+                    : null;
+                string actionName = routeValues.TryGetValue("action", out actionName)
+                    ? actionName?.ToLowerInvariant()
+                    : null;
+                string areaName = routeValues.TryGetValue("area", out areaName)
+                    ? areaName?.ToLowerInvariant()
+                    : null;
+                string pagePath = routeValues.TryGetValue("page", out pagePath)
+                    ? pagePath?.ToLowerInvariant()
+                    : null;
+                string aspNetRoute = trackingFeature.Route;
+                string resourceName = trackingFeature.ResourceName;
+
+                if (aspNetRoute is null || resourceName is null)
+                {
+                    // Not using endpoint routing
+                    string rawRouteTemplate = actionDescriptor.AttributeRouteInfo?.Template;
+                    RouteTemplate routeTemplate = null;
+                    if (rawRouteTemplate is not null)
+                    {
+                        try
+                        {
+                            routeTemplate = TemplateParser.Parse(rawRouteTemplate);
+                        }
+                        catch { }
+                    }
 
-                string httpMethod = request.Method?.ToUpperInvariant() ?? "UNKNOWN";
-                string routeTemplate = actionDescriptor.AttributeRouteInfo?.Template;
-                if (routeTemplate is null)
+                    if (routeTemplate is null)
+                    {
+                        var routeData = httpContext.Features.Get<IRoutingFeature>()?.RouteData;
+                        if (routeData is not null)
+                        {
+                            var route = routeData.Routers.OfType<RouteBase>().FirstOrDefault();
+                            routeTemplate = route?.ParsedTemplate;
+                        }
+                    }
+
+                    if (routeTemplate is not null)
+                    {
+                        // If we don't have a route, don't overwrite the existing resource name
```

Comment:
I found this comment a little confusing trying to reason through a double-negative, how about this?
```suggestion
                        // If we have a route, use the route to determine the resource name
```